src/nets/me2e_asr_attctc.py

```diff
@@ -0,0 +1,204 @@
+
+        return logmel_feats
+
+
+class CMVN(chainer.Chain):
+
+    def __init__(self, cmvn_stats):
+        super(CMVN, self).__init__()
+
+        self.cmvn_mean = cmvn_stats[0]
+        self.cmvn_std = cmvn_stats[1]
+
+    def __call__(self, feats, xp):
+        cmvn_mean = chainer.Variable(xp.array(self.cmvn_mean, dtype=np.float32))
+        cmvn_std = chainer.Variable(xp.array(self.cmvn_std, dtype=np.float32))
+
+        cmvn_feats = []
+        for feat in feats:
+            feat -= F.broadcast_to(cmvn_mean, shape=feat.shape)
+            feat /= F.broadcast_to(cmvn_std, shape=feat.shape)
+            cmvn_feats.append(feat)
+
+        return cmvn_feats
+
+
+class ME2E(chainer.Chain):
+    def __init__(self, enhan, asr, melmat, cmvn_stats):
+        super(ME2E, self).__init__()
+
+        with self.init_scope():
+            # enhan
+            self.enhan = enhan
+            # asr
+            self.asr = asr
+
+        self.logmel = LogMel(melmat)
+        self.cmvn = CMVN(cmvn_stats)
+
+    # x[i]: ('utt_id', {'ilen':'xxx',...}})
+    def __call__(self, data):
+        '''ME2E forward
+
+        :param data:
+        :return:
+        '''
+        mode = data[0][1]['mode']
+        logging.info('mode: ' + mode)
+
+        if mode == 'noisy':
+            # utt list of frame x dim
+            xs_real = [i[1]['feat']['real'] for i in data]
+            xs_imag = [i[1]['feat']['imag'] for i in data]
+            ilens = self.xp.array([utt.shape[0] for utt in xs_real], dtype=np.int32)
+            hs_real = [chainer.Variable(self.xp.array(utt, dtype=np.float32)) for utt in xs_real]
+            hs_imag = [chainer.Variable(self.xp.array(utt, dtype=np.float32)) for utt in xs_imag]
+
+            hs = {}
+            hs['real'] = hs_real
+            hs['imag'] = hs_imag
+        elif mode == 'enhan':
+            # utt list of channel list of frame x dim
+            xs_real = [i[1]['feat']['real'] for i in data]
+            xs_imag = [i[1]['feat']['imag'] for i in data]
+            ilens = self.xp.array([utt[0].shape[0] for utt in xs_real], dtype=np.int32)
+            hs_real = [[chainer.Variable(self.xp.array(ch, dtype=np.float32))
+                        for ch in utt] for utt in xs_real]
+            hs_imag = [[chainer.Variable(self.xp.array(ch, dtype=np.float32))
+                        for ch in utt] for utt in xs_imag]
+
+            hs = {}
+            hs['real'] = hs_real
+            hs['imag'] = hs_imag
+
+            # 1. beamformer
+            hs = self.enhan(hs)
+        else:
+            logging.error(
+                "Error: need to specify an appropriate training mode")
+            sys.exit()
+
+        # utt list of olen
+        ys = [self.xp.array(
+            list(map(int, i[1]['tokenid'].split())), dtype=np.int32) for i in data]
+        ys = [chainer.Variable(y) for y in ys]
+
+        # 2. feature extractor
+        hs = self.logmel(hs, self.xp)
+        hs = self.cmvn(hs, self.xp)
+
```

Comment:
If you assume `self.asr.subsample[0] == 1` here, add `assert self.asr.subsample[0] == 1` in `__init__`.

Reply:
or a kind of `NotImplementedError`

Reply:
I have check again, but I think we can conduct subsample in the asr's first layer.
(Although we cannot conduct subsample in beamforming network.)